Add error handling to the API blueprint.

scolby33 opened the pull request:
Will close #25.

scolby33 committed:
Add error handling to /register API endpoint.

Start of work on #25.

scolby33 committed:
Add :raises: to docstrings where appropriate.

scolby33 committed:
Ignore McCabe complexity on register_location_key.

scolby33 committed:
Add more error checking to the API.

scolby33 committed:
Ignore McCabe complexity on submit.

cthoyt commented on src/ocspdash/web/blueprints/api.py:
```diff
@@ -92,26 +115,45 @@ def _prepare_result_dictionary(result_data):
-    prepared_result_dicts = (_prepare_result_dictionary(result_data)
-                             for result_data in claims[OCSP_RESULTS_JWT_CLAIM])
+    try:
+        prepared_result_dicts = (_prepare_result_dictionary(result_data)
```
Come on, use proper spacing :)

scolby33 replied:
What did I do wrong here?

cthoyt commented on src/ocspdash/web/blueprints/api.py:
```diff
@@ -92,26 +115,45 @@ def _prepare_result_dictionary(result_data):
+    except (KeyError, ValueError):
+        abort(HTTPStatus.BAD_REQUEST, 'invalid result data')
+
+    # TODO: can this raise an exception? I think yes if there's a constraint broken on the DB when commit is called()
```
catch an integrity error then

cthoyt commented on src/ocspdash/web/blueprints/api.py:
```diff
@@ -65,7 +85,7 @@ def get_manifest():
     n = request.args.get('n', type=int, default=10)  # TODO make configurable at app level
     if n > 10:
-        abort(400, 'n too large, max is 10')  # TODO get the max config value here too
+        abort(HTTPStatus.BAD_REQUEST, 'n too large, max is 10')  # TODO get the max config value here too
```
is there a wrapper for aborting on 400?

scolby33 replied:
No, the flask docs have things like `abort(404)`. It is annoying to be so much wordier with `HTTPStatus.BAD_REQUEST` but I think that's better than a "magic" number even if most everyone knows it.

cthoyt commented on src/ocspdash/web/blueprints/api.py:
```diff
@@ -25,25 +26,44 @@
+        invite_token = b64decode(claims['token'])
+    except KeyError:
+        abort(HTTPStatus.BAD_REQUEST, "'token' missing from claims")
+    except binascii.Error:
```
that's an awful name for an error

scolby33 replied:
I take what the stdlib provides.

cthoyt replied:
namespaces, those are a honking good thing

cthoyt reviewed (approved):
Overall looks good!

Requires lots more unit testing, though. And having everything go through 400 makes it tricky to pick out that the right error comes at the right time (besides doing string checking)

scolby33 commented:
I agree so many 400's is not the best, but none of the other error codes seem to fit better.

One thing I have seen is returning JSON errors like (for Pushover, for example):

```json
{
  "user": "invalid",
  "errors": [
    "user identifier is invalid"
  ],
  "status": 0,
  "request": "5042853c-402d-4a18-abcb-168734a801de"
}
```

Despite them having a list of `errors`, I don't think I ever saw a response with more than one there.

cthoyt commented:
You can return json and give it a non-200 error

scolby33 commented:
Well, gonna try this now http://flask.pocoo.org/docs/1.0/patterns/apierrors/

scolby33 committed:
Switch to using an Exception instead of manually calling abort()

http://flask.pocoo.org/docs/1.0/patterns/apierrors/

Changes for flake and mypy.

scolby33 committed:
Remove some print() debugging. Sorry @cthoyt

scolby33 committed:
Fancier error messages for the API.

scolby33 committed:
Add `status` key to the response JSON of InvalidUsage exception.

scolby33 committed:
Broken client_session fixture first try.

scolby33 committed:
Style fixes.

scolby33 committed:
Add the ability to do tests on the Flask portion of OCSPdash.

- Add client_session and client_function fixtures to create Flask test clients and rollback the db between tests
- Add get_db staticmethod to OCSPSQLAlchemy class
- Add the ability to pass session_options into the flask-sqlalchemy SQLAlchemy constructor from the create_application function
  so the test fixtures can take control of rollbacks.
- Add stub test to start using the test client.

scolby33 committed:
Since this static method is kinda generic, use a more generic return type annotation.

scolby33 commented on tests/conftest.py:
```diff
@@ -135,12 +135,86 @@ def manager_function(manager_session):
+    logger.debug('rolling back transaction from function for web client')
+    transaction.rollback()
+
 
```
@cthoyt if you can figure out how to reduce the repetition between `manager_session` and `client_session` and between `manager_function` and `client_function` that'd be swell.

But that's not your real homework.

scolby33 commented on tests/conftest.py:
```diff
@@ -135,12 +135,86 @@ def manager_function(manager_session):
+    transaction.rollback()
+
 
+# TODO: fixture to pre-fill DB with some stuff for the client to test on
```
@cthoyt this is the real homework.

Something that takes in a client and spits out the same client with a db with a bunch of stuff in it.

It'd be cool if it *didn't* get rolled back, but that might be a huge pain so whatever works.

cthoyt committed:
Update type hints and documentation

Scoped sessions act like sessions... why don't they just inherit from them?!?

cthoyt committed:
Remove unnecessary default

The SQLAlchemy class takes a none, so you can pass it none (unless i'm an idiot and none means something special)

cthoyt committed:
Combine logic

cthoyt committed:
Add function to make some test data

@scolby33 not sure where you want this, but here's the idea with making some test data

scolby33 commented on src/ocspdash/manager.py:
```diff
@@ -524,11 +525,11 @@ def process_location(self, invite_token: bytes, public_key: str) -> Optional[Loc
         location = self.get_location_by_selector(selector)
         if location is None:
-            raise Exception(f'location not found for selector: {selector}')
+            raise ValueError(f'invalid invite token')
```
Are these [vexing exceptions](https://ericlippert.com/2008/09/10/vexing-exceptions/)? @cthoyt

cthoyt replied:
No, since it doesn't mean your code has been designed wrong, but rather you have to deal with the reality that someone else is going to stick weird shit inside as the invite token.

Now, the next exception checking for the expiration of the pubkey might be different, but raising an exception actually makes the logic for using this function much better. Overall, I like this design and I think it does what it's supposed to in an elegant way. Lots of helpful/useful errors for bad situations where the program should stop seems good to me.

cthoyt commented on src/ocspdash/web/blueprints/api.py:
```diff
@@ -30,40 +31,41 @@
-    except jwt.JWTError:
-        abort(HTTPStatus.BAD_REQUEST, 'malformed JWT')
+    except jwt.JWTError as e:
+        raise InvalidUsage(f'failed to decode JWT: {str(e)}')
```
you could also do 
```python
raise InvalidUsage('failed to decode JWT') from e
```
unless the point of this was to get all of that information in the string

cthoyt commented on src/ocspdash/web/blueprints/api.py:
```diff
@@ -30,40 +31,41 @@
+    except ValueError as e:
+        raise InvalidUsage(f'failed to process invite: {str(e)}')
 
     return '', HTTPStatus.NO_CONTENT
```
I still don't like returning no content... why don't you have it send back some information about the location that's been registered?

scolby33 replied:
201 Created specifies you should return the URL of the created entity but there is no URL for the location.

What would you want to return? The location name? I was originally just jsonifying the location, but there's some fields that probably shouldn't be returned and going down the road of to_json and safe_to_json screams bad to me.

cthoyt committed:
Address TODOs

- Use app configuration for default and maximum manifest size
- Handle integrity error when inserting payload
- Also add raising of integrity errors to documentation
- make secret key more secret

cthoyt commented:
yo whats going on with this PR

cthoyt commented:
BUMP
